{"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.10.12"}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"markdown","source":"# Bio722P Coding for Bioscientists Final Assignment, 2022/23","metadata":{}},{"cell_type":"markdown","source":"This assignment counts towards 100% of the marks for this module. The assignment is divided into four questions, each carrying 25 marks.","metadata":{}},{"cell_type":"markdown","source":"## Student name and number\nEnter your full name and student number in the cell below","metadata":{}},{"cell_type":"markdown","source":"Full name: Fahmida Tanzeem\n\nstudent number: 150427049","metadata":{}},{"cell_type":"markdown","source":"\n## Deadline and submission\n\nComplete this notebook with your code, compress it and submit it as a single .zip or .gz file through the QMPlus page by the following deadline:\n\n    Friday Nov 3rd 2023, 16:00\n\n\n## Submission checklist\n\n* Notebook is written in Python 3 and can be opened and run on the EECS/SBCS JupyterHub.\n* Any import statements are grouped in the cell provided at the end of this introduction; name and version of the library are given in a comment. There should be no other import statements in the notebook.\n* Notebook functions pass all tests when cells are run with a fresh kernel from top to bottom.\n* Notebook only contains well commented answer code and the original tests. Please remove your own test code.\n* Notebook includes the full name and student number of the author.\n* Zip file containing the notebook is free from errors.\n\nPlease refrain from editing the text of the assignment and the test code provided. Feel free to insert text or code cells as needed, but please clean up rough work that you do not wish to be marked.\n\n## Marking\n\nYour work will be marked **automatically**. Marks will be based on:\n\n        completeness and correctness: 100%\n        \nFor automatic marking to function appropriately, each of your answers **must** pass the tests provided when running the notebook from top to bottom on a fresh kernel. Code that does not pass the tests or is not run (directly or indirectly, e.g. a function calling another function) by the functions you are asked to implement will **not** be marked.\n\nThe test code provided with each question does **not** ensure that your submission is correct; it only ensures that it can be marked automatically. You will need to test your code independently for correctness; however, please remove your test code from your final submission, and do not submit any test data; the marking script will test your code with its own choice of data.\n\n## How to answer questions\n\nThis assignment contains four questions, that can in principle be tackled in any order. However, we suggest that you answer Question 3 before attempting Question 4.\n\nEach question will require you to code a few functions. As you progress through the subquestions, you may have to rely on your answer to previous subquestions to perform some operations. Please call the relevant functions and avoid duplicating code.  \n\n\n## Use of libraries\n\nThis assignment is designed to be feasible using only the functions in the Python standard library, and we recommend that you do not use anything else. However, you are free to use any library available in PyPI if you think that would make your work easier. If you do decide to use any additional libraries, these must be installable from PyPI using *pip*. Specify in the cell below all necessary import statements; for functions that are not part of the standard library, include the name and version of the library in a comment.","metadata":{}},{"cell_type":"code","source":"# Please put all your import statements here, according to the example. Include a comment with the name of \n# the library and its version. Do not include import statements in the rest of the code\n\n# Example:\n# Biopython v 1.78\n# import Bio # this would obviously be uncommented if you are importing Biopython","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Preliminary reading\n\nPlease read the PDF file ```BIO723P_Assignment_Background.pdf``` distributed with this assigment carefully before you start coding.","metadata":{}},{"cell_type":"markdown","source":"# Question 1: Primers and melting temperature","metadata":{}},{"cell_type":"markdown","source":"### Question 1.a: Reading a DNA sequence from a file\n\nWrite a python function called ```readDNAsequence``` that takes as its argument the name of a file.  When passed the name of a FASTA file, the function should read the file, discard the header and return the sequence as a string. Your code should raise ```BadSequenceException``` (defined below) if the sequence part of the file contains characters that are not one of the letters A, C, T, G, U. All U nucleotides should be replaced by T in the returned string (for simplicity, we will be working with T only throughout the rest of this assignment).","metadata":{}},{"cell_type":"code","source":"# Run this cell to define the exception\nclass BadSequenceException(Exception):\n    pass","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Your code here\nclass BadSequenceException(Exception):\n    pass\n\ndef readDNAsequence(filename):\n    with open(filename, 'r') as f:\n        lines = f.readlines()\n\n    sequence_lines = lines[1:]\n\n#  lines Joined into a single string and newlines removed\n    sequence = ''.join(sequence_lines).replace('\\n', '')\n\n    #  invalid characters Check\n    for char in sequence:\n        if char not in 'ACTGU':\n            raise BadSequenceException(\"Invalid character found\")\n\n    # 'U' with 'T' Replaced\n    sequence = sequence.replace('U', 'T')\n\n    return sequence\n","metadata":{"trusted":true,"tags":[]},"execution_count":6,"outputs":[]},{"cell_type":"code","source":"# Test code\nimport os\nwith open(\"test9876345.fas\", \"wt\") as _OUTF:\n    _OUTF.write(\"> test\\n\")\n    _OUTF.write(\"ACTG\\n\")\n_seq=readDNAsequence(\"test9876345.fas\")\nassert type(_seq) is type(\"\"), \"Return value is not a string: %r\" % _seq\nos.remove(\"test9876345.fas\")\nprint(\"OK\")","metadata":{"trusted":true,"tags":[]},"execution_count":4,"outputs":[{"name":"stdout","text":"OK\n","output_type":"stream"}]},{"cell_type":"markdown","source":"### Question 1.b: Computing the complement of a sequence","metadata":{}},{"cell_type":"markdown","source":"Write a function called ```complement``` that takes a string containing a DNA sequence as its only  parameter and returns the complement of the sequence in a string. The function should raise ```BadSequenceException``` if the argument sequence contains anything else than the four characters A, C, T, G. Do not reverse the string; for the avoidance of doubt, if the input string starts with A then the complement string should start with T.\n","metadata":{}},{"cell_type":"code","source":"# Your code here\n# To define the exception\nclass BadSequenceException(Exception):\n    pass\n\n# Function to find the complement of DNA sequence\ndef complement(dna_sequence):\n    complement_dict = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C'}\n    complement_sequence = ''\n\n    for char in dna_sequence:\n        if char not in \"ACTG\":\n            raise BadSequenceException(\"Invalid character in sequence\")\n        complement_sequence += complement_dict[char]\n\n    return complement_sequence","metadata":{"trusted":true,"tags":[]},"execution_count":9,"outputs":[]},{"cell_type":"code","source":"# Test code\n_seq=complement(\"ACGTTTCG\")\nassert type(_seq) is type(\"\"), \"Return value is not a string: %r\" % _seq\nprint(\"OK\")","metadata":{"trusted":true,"tags":[]},"execution_count":10,"outputs":[{"name":"stdout","text":"OK\n","output_type":"stream"}]},{"cell_type":"markdown","source":"### Question 1.c: Extracting primers","metadata":{}},{"cell_type":"markdown","source":"Write a function called ```primer``` that takes three parameters: a DNA sequence called ```sequence```, an integer ```length``` that is 20 by default and a Boolean value ```forward``` that is ```True``` by default. When ```forward``` is ```True``` (or is not passed), the function should return a Forward primer for the sequence passed as ```sequence```; when it is ```False```, it should return a  Reverse primer. The length of the primer is specified by ```length```; if this is not passed, a primer of length 20 should be returned.  Refer to the Background document for a definition of primers and how to compute them (for the avoidance of doubt, if the sequence string ends with a C, then the reverse primer string should start with a G). If the sequence is shorter than ```length``` nucleotides, your code should raise a ```BadSequenceException```.\n","metadata":{}},{"cell_type":"code","source":"# Your code here\n# To define the exception\nclass BadSequenceException(Exception):\n    pass\n\n# Function to find primers\ndef primer(sequence, length=20, forward=True):\n    if len(sequence) < length:\n        raise BadSequenceException(\"Sequence shorter than specified length\")\n\n    if forward:\n        return sequence[:length]\n    else:\n        complement_dict = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C'}\n        reverse_primer = ''.join(complement_dict[char] for char in reversed(sequence[-length:]))\n        return reverse_primer","metadata":{"trusted":true,"tags":[]},"execution_count":1,"outputs":[]},{"cell_type":"code","source":"# Test code\n_seq=primer(sequence=\"AAAAATTTTTCCCCCGGGGGAAAAA\", length= 20, forward=False)\nassert type(_seq) is type(\"\"), \"Return value is not a string: %r\" % _seq\nprint(\"OK\")","metadata":{"trusted":true,"tags":[]},"execution_count":2,"outputs":[{"name":"stdout","text":"OK\n","output_type":"stream"}]},{"cell_type":"markdown","source":"### Question 1.d: Computing the melting temperature","metadata":{}},{"cell_type":"markdown","source":"Write a function called ```meltingTemp``` that takes a string representing a primer as its argument. The function should return the melting temperature of the primer in degrees Celsius according to the equation given in the Background document. If the sequence contains characters other than A, C, T, G, the function should raise a ```BadSequenceException```.","metadata":{}},{"cell_type":"code","source":"# Your code here\n# To define the exception\nclass BadSequenceException(Exception):\n    pass\n\n# Function to calculate melting temperature\ndef meltingTemp(primer):\n    for char in primer:\n        if char not in \"ACTG\":\n            raise BadSequenceException(\"Invalid character in sequence\")\n\n    A_T_count = primer.count('A') + primer.count('T')\n    C_G_count = primer.count('C') + primer.count('G')\n\n    melting_temp = (2 * A_T_count) + (4 * C_G_count)\n    return melting_temp\n","metadata":{"trusted":true,"tags":[]},"execution_count":11,"outputs":[]},{"cell_type":"code","source":"# Test code\n_temp=meltingTemp(\"AAAAATTTTTCCCCCGGGGG\")\nassert ((type(_temp) is type(0.0)) or\n        (type(_temp) is type(0))), \"Return value is not a number: %r\" % _temp\nprint(\"OK\")","metadata":{"trusted":true,"tags":[]},"execution_count":14,"outputs":[{"name":"stdout","text":"OK\n","output_type":"stream"}]},{"cell_type":"markdown","source":"### Question 1.e: Putting it all together\n\nWrite a function called ```sequencePCRtemp``` that takes a string containing the name of a FASTA file as its argument. The function should return the average melting temperature of the two primers of the sequence as a ```float```.","metadata":{}},{"cell_type":"code","source":"# Your code here\n# To define the exception\nclass BadSequenceException(Exception):\n    pass\n\n# Your existing functions for reading DNA, complement, and melting temp should be here\n\ndef sequencePCRtemp(file_name):\n    try:\n        # Read the DNA sequence from the file\n        dna_sequence = readDNAsequence(file_name)\n\n        # Extract forward and reverse primers\n        forward_primer = primer(dna_sequence)\n        reverse_primer = primer(dna_sequence, forward=False)\n\n        # Calculate melting temperatures for both\n        melting_temp_forward = meltingTemp(forward_primer)\n        melting_temp_reverse = meltingTemp(reverse_primer)\n\n        # Calculate and return the average melting temperature\n        average_temp = (melting_temp_forward + melting_temp_reverse) / 2\n        return average_temp\n\n    except FileNotFoundError:\n        raise FileNotFoundError(f\"File '{file_name}' not found\")\n    except BadSequenceException as e:\n        raise e\n","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Test code\nimport os\nwith open(\"test9876346.fas\", \"wt\") as _OUTF:\n    _OUTF.write(\"> test\\n\")\n    _OUTF.write(\"AAAAACCCCCTTTTTGGGGGAAAAA\\n\")\n_temp=sequencePCRtemp(\"test9876346.fas\")\nassert type(_temp) is type(0.0), \"Return value is not a float: %r\" % _temp\nos.remove(\"test9876346.fas\")\nprint(\"OK\")","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Question 2: Translation and reading frames","metadata":{}},{"cell_type":"markdown","source":"### Question 2.a: Reading frames\nWrite a function ```translate``` that takes a string containing a DNA sequence as its input and outputs a Python dictionary containing the translation of the sequence in all possible reading frames. The keys of the dictionary should be ```f1```, ```f2```, ```f3``` for the three forward frames and ```r1```, ```r2``` and ```r3``` for the reverse reading frames; the value of each key should be the translation of the sequence in the corresponding frame.\nFor simplicity and ease of debugging, **do not complement the sequence** when computing the reverse reading frames; just reverse it. Use an asterisk (```*```) to represent stop codons. Always translate the entire sequence.","metadata":{}},{"cell_type":"code","source":"# Your code here\n","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Test code\n_seqdic=translate(\"ACTGACTGACTGACTGACTGACTG\")\nassert type(_seqdic)==type(dict()), \"Return value is not a dictionary: %r\" % _seqdic\nassert set(_seqdic.keys())==set(['f1', 'f2', 'f3', 'r1', 'r2', 'r3']), \\\n    \"Output dictionary has incorrect/missing keys: %r\"  % _seqdic.keys()\nassert type(_seqdic['f1'])==type(\"\"), \\\n    \"Output dictionary values should be strings, not %r\" % type(_seqdic['f1'])\nprint(\"OK\")","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"### Question 2.b: Locating an ORF\n\nWrite a function called ```openReadingFrame``` that takes a string containing an aminoacid sequence as its argument and returns a string containing the aminoacids between the first Methionine (included) and the first STOP codon that follows it (excluded). Assume the stop codon is represented by an asterisk (```*```) as would be returned by ```translate``` above. If either the Methionine or the STOP codon are missing, your function should return an empty string.","metadata":{}},{"cell_type":"code","source":"# Your code here\n","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Test code\n_seq=openReadingFrame(\"AMCAPP*L\")\nassert type(_seq) is type(\"\"), \"Return value is not a string: %r\" % _seq\nprint(\"OK\")","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"### Question 2.c: Translating a sequence\n\nWrite a function called ```candidateProtein``` that takes a string containing a DNA sequence as its input and outputs the string of aminoacids corresponding to the longest ORF, as extracted by ```openReadingFrame``` above.","metadata":{}},{"cell_type":"code","source":"# Test code\n_seq=candidateProtein(\"ATGACTGCTGGGTAG\")\nassert type(_seq) is type(\"\"), \"Return value is not a string: %r\" % _seq\nprint(\"OK\")","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"### Question 2.d: Writing a FASTA file\n\nWrite a function called ```writeFASTA``` that takes three string arguments called, in the order, ```sequence```, ```description``` and ```filename```. Argument ```sequence``` should contain an aminoacid sequence. Argument ```description``` should contain a description (eg name of protein, organism, etc). Argument ```filename``` should contain a file name. Your code should create the file with the name requested, write to  it the description as a FASTA header (i.e. starting with the character ```>```) and write the sequence to the file. Long sequences should be formatted over several lines. The function should not return any value.","metadata":{}},{"cell_type":"code","source":"# Your code here\n","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Test code\nimport os\nimport os.path\n_rv=writeFASTA(sequence=\"TESTTESTTESTTESTTEST\",\n              description=\"test sequence\",\n              filename=\"test9876347.fas\")\nassert type(_rv) is type(None), \"Function should not return anything; it returns %r\" % _rv\n_fe=os.path.isfile(\"test9876347.fas\")\nassert _fe, \"Cannot find output file - has it been created?\"\nos.remove(\"test9876347.fas\")\nprint(\"OK\")","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"### Question 2.e: Putting it all together\n\n\nWrite a function called ```maximalORF``` that takes as its argument string ```inputfile``` containing the name of an input file, string ```outputfile``` with the name of an output file and string  ```proteinname``` with a description of a candidate protein. The function should read a DNA sequence from the input file and write the candidate protein corresponding to the longest ORF to the output file, in FASTA format. The string supplied in ```proteinname``` should provide the header of the FASTA file. The function should not return any value.","metadata":{}},{"cell_type":"code","source":"# Your code here\n","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Test code\nimport os\nimport os.path\nwith open(\"test9876348.fas\", \"wt\") as _OUTF:\n    _OUTF.write(\"> test\\n\")\n    _OUTF.write(\"ATGACTGCTGGGTAG\\n\")\n_rv=maximalORF(inputfile=\"test9876348.fas\", outputfile=\"test9876349.fas\",\n               proteinname=\"test protein\")\nassert type(_rv) is type(None), \"Function should not return anything; it returns %r\" % _rv\n_fe=os.path.isfile(\"test9876349.fas\")\nassert _fe, \"Cannot find output file - has it been created?\"\nos.remove(\"test9876348.fas\")\nos.remove(\"test9876349.fas\")\nprint(\"OK\")","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Question 3: Classification of aminoacids","metadata":{}},{"cell_type":"markdown","source":"### Question 3.a: Reading an aminoacid sequence\n\nWrite a function called ```readAAsequence``` that takes as its argument the name of a file.  When passed the name of a FASTA file containing an aminoacid sequence, the function should read the file, discard the header and return the sequence as a string.","metadata":{}},{"cell_type":"code","source":"# Your code here\n","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Test code\nimport os\nwith open(\"test9876350.fas\", \"wt\") as _OUTF:\n    _OUTF.write(\"> test\\n\")\n    _OUTF.write(\"TESTTESTTESTTESTTEST\\n\")\n_seq=readAAsequence(\"test9876350.fas\")\nos.remove(\"test9876350.fas\")\nassert type(_seq) is type(\"\"), \"Return value is not a string: %r\" % _seq\nprint (\"OK\")","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"### Question 3.b: Aminoacid usage statistics\n\nWrite a function called ```AAtypes``` that takes as its argument a string containing a sequence of aminoacids. The function should compute, for each protein, the fraction of aminoacids that are polar, small and hydrophobic and return that as a list. For instance, if the chain of a protein contains 35% polar residues, 15% small and 50% hydrophobic residues, the function should return ```[0.35, 0.15, 0.5]``` (the order is important). Refer to the Venn diagram in the Assessment Background file for the classification of aminoacids (other Venn diagrams may differ).  Note that some aminoacids belong to more than one category, so that the sum of these three numbers can be larger than 1.","metadata":{}},{"cell_type":"code","source":"# Your code here\n","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Test code\n_aatypes=AAtypes(\"TESTTESTTESTTESTTEST\")\nassert type(_aatypes) is type([]), \"Return value is not a list: %r\" % _aatypes\nassert len(_aatypes)==3, \"Returned list should contain 3 values\"\nprint(\"OK\")","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"### Question 3.c: Processing multiple sequences\nWrite a function called ```AAtypetable``` that takes a list of strings called ```filelist``` and a separate string called ```outputfile```. Each element of ```filelist``` represents the name of a FASTA file containing an aminoacid sequence. For each filename, the function should load the sequence, compute the fraction of aminoacids that are polar, small and/or hydrophobic, and finally output this to the text file specified by ```outputfile``` in TSV (tab-separated values) tabular format. The table should contain, on each line, the name of the input file and then the percentages of polar, small and hydrophobic residues (in that order). Include a comment at the beginning of the file that starts with a ```#``` and specifies the content of each column, as follows (numbers and file names shown here are random):\n```\n# Filename     Polar     Small     Hydro\nPxxx.fasta     0.52      0.22      0.33\nQx3Z.fas       0.47      0.35      0.38\n...\n```\n\nInclude at least two decimal places in your output; do not add any quotes or other spurious characters. Do not worry if the columns do not all look aligned; as long as consecutive values on the same line are separated by a tab, that's fine. If some of the files in ```filelist``` do not exist, contain invalid data or otherwise cause an error, your function should ignore those files and continue. However, the names of those files should be returned by your function in a list. If all files are processed without errors your function should return an empty list.\n","metadata":{}},{"cell_type":"code","source":"# Your code here\n","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Test code\nimport os\nimport os.path\nfor fname in [\"_Ptest123.fas\", \"_Ptest456.fas\"]:\n    with open(fname, \"wt\") as _OUTF:\n        _OUTF.write(\"> test\\n\")\n        _OUTF.write(\"TESTTESTTESTTESTTEST\\n\")\n_rv=AAtypetable(filelist=['_Ptest123.fas','_Ptest456.fas'], outputfile=\"_table_test789.txt\")\n_fe=os.path.isfile(\"_table_test789.txt\")\nassert _fe, \"Cannot find output file - has it been created?\"\nos.remove(\"_table_test789.txt\")\nos.remove(\"_Ptest123.fas\")\nos.remove(\"_Ptest456.fas\")\nassert type(_rv) is type([]), \"Function should return a list; it returns %r\" % _rv\nprint(\"OK\")","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Question 4: Clustering proteins based on aminoacid usage","metadata":{}},{"cell_type":"markdown","source":"### Question 4.a: Computing the Euclidean distance","metadata":{}},{"cell_type":"markdown","source":"Write a function called ```distance``` that takes two lists of ```float``` of equal length as its arguments. The function should return the Euclidean distance between the two lists, considered as vectors in a space of dimensionality equal to the length of the lists. That is to say, it should compute the element by element difference between the two lists, square that, sum all the squares and take a square root of the final sum (see the equation [here](https://en.wikipedia.org/wiki/Euclidean_distance#Higher_dimensions)). Your function should raise a ```DimensionalityException``` if the two lists passed are not of the same length, or if they are both empty; otherwise it should return a ```float```.","metadata":{}},{"cell_type":"code","source":"# Run this cell to define the exception\nclass DimensionalityException(Exception):\n    pass","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Your code here\n","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Test code\n_dist=distance([1.0, 2.0], [1.5, -1.0])\nassert type(_dist) is type(0.0), \"Return value is not a float: %r\" % _dist\nprint(\"OK\")","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"### Question 4.b: Reading data in tabular format\n\nWrite a function called ```readTable``` that takes a string called ```filename``` as its argument. The function should read from the file specified in the argument a TSV table formatted as described at the end of Question 3 (you do not necessarily have to answer Question 3 in order to do this; you can create one such table containing arbitrary values using an editor). The function should return a dictionary with the filenames as its keys; the value associated to each key should be a list of floats, containing in order the percentages of polar, small and hydrophobic AAs, as listed in the table. So for instance, if the argument file contains the table used as an example in Question 3, ```readTable``` should return a dictionary containing at least the following items: \n```\n{ \"Pxxx.fasta\": [0.52, 0.22, 0.33],  \"Qx3Z.fas\": [0.47, 0.35, 0.38],\n... }\n```\n","metadata":{}},{"cell_type":"code","source":"# Your code here\n","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Test code\nimport os\nwith open(\"_testTable123.tsv\", \"wt\") as OUTF:\n    OUTF.write(\"# Filename\\tPolar\\tSmall\\tHydro\\n\")\n    OUTF.write(\"Pxxx.fas\\t0.52\\t0.22\\t0.33\\n\")\n    OUTF.write(\"Pyyy.fas\\t0.47\\t0.35\\t0.38\\n\")\n_tab=readTable(filename=\"_testTable123.tsv\")\nassert type(_tab) is type({}), \"Return value is not a dict: %r\" % _tab\nassert (\"Pxxx.fas\" in _tab) and (\"Pyyy.fas\" in _tab), \"Return dict missing keys\"\nos.remove(\"_testTable123.tsv\")\nprint(\"OK\")","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"### Question 4.c: Computing a distance matrix\n\nWrite a function called ```distanceMatrix``` that takes as its argument two strings called ```inputfile``` and ```outputfile```, used to pass the names of the input and output file. The function should read from the input file a TSV table created as specified at the end of Question 3 (again, you do not necessarily have to answer Question 3 in order to do this; you can create a table with made-up data using an editor). Suppose data for N proteins are read. Your function should then create an output TSV file containing a matrix of size NxN (plus one row and one column for labels). The rows and columns of the matrix should correspond to individual protein file names, so that each matrix entry corresponds to a pair of proteins, say Pxxx and Pyyy. Such entry should contain the distance between the ```[polar, small, hydro]``` lists (considered as 3 dimensional vectors) corresponding to Pxxx and Pyyy. If done correctly, your matrix should be symmetric and have zeroes on the diagonal. Format the matrix with row and column labels as in the following 3-protein example (that contains random numbers and filenames):\n```\n# Filename      Pxxx.fas  Pyyy.fas  Pzzz.fas\nPxxx.fas        0.00      2.34      1.51\nPyyy.fas        2.34      0.00      1.82 \nPzzz.fas        1.51      1.82      0.00\n\n```\nThe exact alignment is not important, as long as the data are separated by tabs and the first line starts with a hash (```#```). Likewise, the order in which the proteins are listed is not important, provided the same order is used along the rows and along the columns. Do not add any quotes or other spurious characters. Your function should not return any value.","metadata":{}},{"cell_type":"code","source":"# Your code here\n","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Test code\nimport os\nimport os.path\nwith open(\"_testTable123.tsv\", \"wt\") as OUTF:\n    OUTF.write(\"# Filename\\tPolar\\tSmall\\tHydro\\n\")\n    OUTF.write(\"Pxxx.fas\\t0.52\\t0.22\\t0.33\\n\")\n    OUTF.write(\"Pyyy.fas\\t0.47\\t0.35\\t0.38\\n\")\n_rv=distanceMatrix(inputfile=\"_testTable123.tsv\", outputfile=\"_testTable123out.tsv\")\n_fe=os.path.isfile(\"_testTable123out.tsv\")\nassert _fe, \"Cannot find output file - has it been created?\"\nwith open(\"_testTable123out.tsv\", \"rt\") as INF:\n    _fl=INF.readline()\nassert _fl[0]=='#', \"First line of output file does not start with a #\"\nos.remove(\"_testTable123.tsv\")\nos.remove(\"_testTable123out.tsv\")\nassert type(_rv) is type(None),\"Function should not return any value; it returns %r\" % _rv\nprint(\"OK\")","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"### An application of this type of analysis","metadata":{}},{"cell_type":"markdown","source":"Now that you have this code running, try downloading a few FASTA sequences for the [antifreeze proteins](https://pdb101.rcsb.org/motm/120) of cold water fish. Then download the sequences of other families of related proteins (eg Hemoglobin, Insulin and P53) from other species. Process these sequences according to Questions 3 and 4. Have a look at the resulting distance matrix - is any of these groups of proteins characterised by a distinctive pattern of usage of aminoacids? NOTE: no marks are assigned for this section.","metadata":{}},{"cell_type":"markdown","source":"## Examiners\n\nDr Fabrizio Smeraldi *(f.smeraldi@qmul.ac.uk)*; Prof Conrad Bessant *(c.bessant@qmul.ac.uk)*","metadata":{}},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]}]}
